_posts/2023-07-29-oauth-authentication-scala-http4s.md

```diff
@@ -0,0 +1,672 @@
+The `/index.html` route serves an `index.html` static file to the browser. We use `StaticFile.fromPath()` which takes the path to our file and an `Option` of `request` passed from `case request @ GET`. Finally, we call `getOrElseF(NotFound())` that returns `404` status code in case the file is not found.
+
+```scala
+import org.http4s.dsl.impl.QueryParamDecoderMatcher
+import com.xonal.oAuth.OauthImpl.getOauthResults
+import cats.syntax.all.*
+
+object GithubRoutes{
+  object GithubTokenQueryParamMatcher
+      extends QueryParamDecoderMatcher[String]("code")
+  def githubRoutes[F[_]: Async]: HttpRoutes[F] = {
+      ...
+      case GET -> Root / "callback" :? GithubTokenQueryParamMatcher(code) =>
+        getOauthResults(code).handleError(_.getMessage).flatMap(Ok(_))
+  }
+}
+```
+
+The second route `/callback?code={single-use authorization code}` contains the URL parameter, `code` which we handle with `http4s`. This is done through the use of the `QueryParamDecoderMatcher` class from `org.http4s.dsl.impl` that takes the parameter key in our case `code`, and returns its value.
+
+We apply this class by extending `GithubTokenQueryParamMatcher` and calling it as part of the route, `case GET -> Root / "callback" :? GithubTokenQueryParamMatcher(code)`.
+
+We then call `getOauthResults(code)` which returns an F[String] containing the response from GitHub's API, however in case the function fails for any reason we use the `handleError(_.getMessage)` method to take care of any errors. We conclude by sending the result back to the browser by calling .flatMap(Ok(\_)). Here the result is either the error message or the GitHub server response.
+
+Here's the full code:
+
+```scala
+package com.xonal.routes
+
+import org.http4s.dsl.impl.QueryParamDecoderMatcher
+import org.http4s.*
+import org.http4s.dsl.Http4sDsl
+import com.xonal.oAuth.OauthImpl.getOauthResults
+import cats.effect.Async
+import fs2.io.file.Path
+import cats.syntax.all.*
+
+object GithubRoutes {
+  object GithubTokenQueryParamMatcher
+      extends QueryParamDecoderMatcher[String]("code")
+
+  def githubRoutes[F[_]: Async]: HttpRoutes[F] = {
+    val dsl = Http4sDsl[F]
+    import dsl.{Path as _, *}
+    HttpRoutes.of[F] {
+      case request @ GET -> Root / "index.html" =>
+        StaticFile
+          .fromPath(
+            Path("oauth/src/main/scala/com/xonal/index.html"),
+            Some(request)
+          )
+          .getOrElseF(NotFound()) // In case the file doesn't exist
+      case GET -> Root / "callback" :? GithubTokenQueryParamMatcher(code) =>
+        getOauthResults(code).handleError(_.getMessage).flatMap(Ok(_))
+    }
+  }
+}
+
+```
+
+Now let's create our index.html file in the following path, `oauth/src/main/scala/com/xonal/index.html` and paste the following code.
+
+```html
+<html>
+  <p>
+    Click
+    <a
+      href="https://github.com/login/oauth/authorize?scope=user:email&client_id=0131cd2e9eacae1854c5"
```

Comment:
instead of your client_id, put something like `client_id=YOUR_CLIENT_ID_HERE`

Reply:
Updated